### PR TITLE
Add metadata to BatchInbox contract

### PR DESCRIPTION
**Pull request by sveitser**

### CODE REVIEW

File: packages/contracts-bedrock/src/L1/LeaderElectionBatchInbox.sol [hunk 1]
```diff
@@ -5,6 +5,40 @@ pragma solidity ^0.8.0;
...
+    ///         encoding for the frames, without having to RLP-decode the frames.
+    ///
+    ///         Checking that the metadata matches the frames must be done in the derivation pipeline. If there is
+    ///         a mismatch between the metadata and the frames, the frames must be discarded.
+    ///
+    ///         Implementations of this contract interface can enforce behaviour based on the metadata submitted
+    ///         to this function.
+    ///
+    /// @param _metas metadata for the frames.
+    function submit(MetaData[] memory _metas) public view {
```
Comment (nyospe) [inline packages/contracts-bedrock/src/L1/LeaderElectionBatchInbox.sol]: This should be `external`, not `public`, and absolutely not `view` (I removed the `view` in [PR 48](https://github.com/EspressoSystems/op-leader-election/pull/48)).

 We're never going to call `submit` from inside another contract.
 
 And then we can change the signature to 
 ```solidity
 function submit(MetaData[] memory _metas, bytes calldata _batch) external {
 ```
 which will not copy the calldata into memory.
 
 It may be possible to keep the metadata as calldata and copy it, bytewise, into the series of needed locals, but I'm not sure if this saves us anything.

File: packages/contracts-bedrock/test/RoundRobinLeaderElection.t.sol [hunk 2]
```diff
@@ -63,11 +63,12 @@ contract RoundRobinLeaderElectionTest is Test {
         address notALeader = vm.addr(1234);
         vm.prank(notALeader);
         vm.expectRevert("RoundRobinLeaderElection: submit function must be called by the leader.");
-        leaderContract.submit("hi");
+        LeaderElectionBatchInbox.MetaData[] memory metas = new LeaderElectionBatchInbox.MetaData[](1);
+        leaderContract.submit(metas);
```
Comment (nyospe) [inline packages/contracts-bedrock/test/RoundRobinLeaderElection.t.sol]: I'm not sure if this works with the `external` specifier... if it doesn't, I'm not sure it's worth testing this way, given the difference between `external` and `public`.

Comment (sveitser) [reply packages/contracts-bedrock/test/RoundRobinLeaderElection.t.sol]: I think it should work.

File: packages/contracts-bedrock/src/L1/LeaderElectionBatchInbox.sol (hunk 1 above)
Comment (sveitser) [reply packages/contracts-bedrock/src/L1/LeaderElectionBatchInbox.sol]: Based on our discussion earlier I thought we wanted to remove the `calldata` argument to avoid RLP encoding/decoding, but I think it's fine either way and having the second parameter feels less hacky.

Agreed that it won't be loaded into memory anyway if it's declared as `calldata` and not used in the function.

If the data is loaded into memory it's cheaper to declare it as memory I think so if it's being used in the function declaring it as memory should be better.

File: packages/contracts-bedrock/test/RoundRobinLeaderElection.t.sol (hunk 2 above)
Comment (sveitser) [reply packages/contracts-bedrock/test/RoundRobinLeaderElection.t.sol]: Yes, confirmed. Since the test contract is another contract that makes sense.

File: packages/contracts-bedrock/src/L1/LeaderElectionBatchInbox.sol [hunk 3]
```diff
@@ -5,6 +5,40 @@ pragma solidity ^0.8.0;
 /// @notice Interface for implementing a leader election scheme
 
 abstract contract LeaderElectionBatchInbox {
+    enum LeaderStatusFlags {
+        Scheduled,
+        Unscheduled,
+        Invalid
+    }
+
+    struct MetaData {
```
Comment (philippecamacho) [inline packages/contracts-bedrock/src/L1/LeaderElectionBatchInbox.sol]: No need to do it for now, by we may add a comment about the need for some commitment/hash to the actual data stored somewhere else, so that the derivation pipeline can check the consistency between metadata and data.

Comment (sveitser) [reply packages/contracts-bedrock/src/L1/LeaderElectionBatchInbox.sol]: I added some comments to the submit function. Does this cover what you had in mind?

https://github.com/EspressoSystems/op-leader-election/blob/61df6481a82cd324aac9bfcdb2fb007c475418cf/packages/contracts-bedrock/src/L1/LeaderElectionBatchInbox.sol#L26-L27

Comment (philippecamacho) [reply packages/contracts-bedrock/src/L1/LeaderElectionBatchInbox.sol]: Yes, your phrasing is more general. Thanks.

File: op-node/rollup/derive/frame.go [hunk 4]
```diff
@@ -130,6 +133,29 @@ func ParseFrames(data []byte) ([]Frame, error) {
 	if len(data) == 0 {
 		return nil, errors.New("data array must not be empty")
 	}
+	// If the calldata starts with the function signature of the submit method
+	// then it is a v2 frame.
+	// TODO: don't parse the ABI every time
```
Comment (nyospe) [inline op-node/rollup/derive/frame.go]: There's an example of caching the ABI in the batcher, on the op_batcher specific Config.

File: op-node/rollup/derive/frame.go [hunk 5]
```diff
@@ -130,6 +133,29 @@ func ParseFrames(data []byte) ([]Frame, error) {
...
+	isV2 := len(data) >= 4 && bytes.Equal(data[:4], submit.ID)
+	var metas []bindings.LeaderElectionBatchInboxMeta
+	if isV2 {
+		metas, data, err = ParseFramesV2(data, submit)
+		if err != nil {
+			return nil, fmt.Errorf("parsing v2 frames: %w", err)
+		}
+	}
+
+	// Otherwise it is a v0 frame.
```
Comment (nyospe) [inline op-node/rollup/derive/frame.go]: This v0 is not related to the V1/V2 on the Batch.

We're still DerivationVersion0. Which seems to be consistent with the code below, but this comment is misleading.

File: packages/contracts-bedrock/src/L1/LeaderElectionBatchInbox.sol [hunk 6]
```diff
@@ -5,6 +5,38 @@ pragma solidity ^0.8.0;
...
+    ///         Checking that the metadata matches the frames must be done in the derivation pipeline. If there is
+    ///         a mismatch between the metadata and the frames, the frames must be discarded.
+    ///
+    ///         Implementations of this contract interface can enforce behaviour based on the metadata submitted
+    ///         to this function.
+    ///
+    /// @param _metas metadata for the frames.
+    /// @param _frames frames to be submitted.
+    function submit(Meta[] memory _metas, bytes calldata _frames) external {
+        // TODO: pass metadata to isCurrentLeader and nextBlockAsLeader
```
Comment (nyospe) [inline packages/contracts-bedrock/src/L1/LeaderElectionBatchInbox.sol]: I don't think `submit` will ever call `nextBlockAsLeader`. That's just for the batchers to use. It does need to get passed to isCurrentLeader.

**Review by nyospe — APPROVED** [review]
A couple of comments that could use some cleanup, but the code looks right.

File: op-node/rollup/derive/frame.go (hunk 5 above)
Comment (sveitser) [reply op-node/rollup/derive/frame.go]: Thanks. Updated the comments.

File: op-node/rollup/derive/frame.go (hunk 4 above)
Comment (sveitser) [reply op-node/rollup/derive/frame.go]: I went with a module level global for the ABI. I think the ABI should always be available and if it isn't the program exiting early is not a bad idea.

File: packages/contracts-bedrock/src/L1/LeaderElectionBatchInbox.sol (hunk 6 above)
Comment (sveitser) [reply packages/contracts-bedrock/src/L1/LeaderElectionBatchInbox.sol]: I updated the comment. I think we can do the actual change to the function once we have a better idea of what should be passed on to the virtual function. Unless anyone thinks we can make the decision already

cc @nyospe @philippecamacho

**Review by philippecamacho — APPROVED** [review]
LGTM
